### DISCUSSION TITLE
Question: How to debug a contract on era-compiler-llvm

### DISCUSSION THREAD

**Discussion by ThreeAndTwo** [zkSync-Community-Hub/zksync-developers#18] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

 I am trying to deploy the Uniswap V2 code on the zkSync ERC local net.

### What is the specific issue or error you're encountering?

An error was encountered when running the test case 'moves the price to 1:90'. However, another one 'moves the price to 1:110' was successful. The specific error information is as follows: 1) ExampleSwapToPrice
#swapToPrice
moves the price to 1:90:
Error: transaction failed [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (transactionHash="0xc129dd7c33e43aa89b0e10f08f23ee70cbe010a240b51bd620347b310fbe60a6", transaction={"nonce":106,"gasPrice":{"type":"BigNumber","hex":"0x11e1a300"},"gasLimit":{"type":"BigNumber","hex":"0xd693a400"},"to":"0x748fA9975Afd2076BAc17B4223be5F3271cc9AC5","value":{"type":"BigNumber","hex":"0x00"},"data":"0xbef90fd300000000000000000000000049dc9fa6e1090a76beb6199460d3179cb6150eb00000000000000000000000005b66afaff2b28a3341a01b429a4648040e16120c0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000005affffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000e90e12261ccb0f3f7976ae611a29e84a6a85f424ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff","chainId":270,"v":575,"r":"0x2918373e82e6af775276938fe80054cbe071801b2b898aebf70b2952143e5134","s":"0x4cb63884e8b7cd5270824608184d3691c0627a44cd627e71de7dcea7f134f6b7","from":"0xE90E12261CCb0F3F7976Ae611A29e84a6A85f424","hash":"0xc129dd7c33e43aa89b0e10f08f23ee70cbe010a240b51bd620347b310fbe60a6","type":null,"confirmations":0}, receipt={"to":"0x748fA9975Afd2076BAc17B4223be5F3271cc9AC5","from":"0xE90E12261CCb0F3F7976Ae611A29e84a6A85f424","contractAddress":null,"transactionIndex":2,"root":"0x9ae51096a60f1704ed381854a01304833bae70d0069aa4312ead2c2d7ee84b28","gasUsed":{"type":"BigNumber","hex":"0x0940e9"},"logsBloom":"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

### Can you share the error messages or logs you're receiving, if any?

Error: transaction failed [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (transactionHash="0xc129dd7c33e43aa89b0e10f08f23ee70cbe010a240b51bd620347b310fbe60a6", transaction={"nonce":106,"gasPrice":{"type":"BigNumber","hex":"0x11e1a300"},"gasLimit":{"type":"BigNumber","hex":"0xd693a400"},"to":"0x748fA9975Afd2076BAc17B4223be5F3271cc9AC5","value":{"type":"BigNumber","hex":"0x00"},"data":"0xbef90fd300000000000000000000000049dc9fa6e1090a76beb6199460d3179cb6150eb00000000000000000000000005b66afaff2b28a3341a01b429a4648040e16120c0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000005affffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff000000000000000000000000e90e12261ccb0f3f7976ae611a29e84a6a85f424ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff","chainId":270,"v":575,"r":"0x2918373e82e6af775276938fe80054cbe071801b2b898aebf70b2952143e5134","s":"0x4cb63884e8b7cd5270824608184d3691c0627a44cd627e71de7dcea7f134f6b7","from":"0xE90E12261CCb0F3F7976Ae611A29e84a6A85f424","hash":"0xc129dd7c33e43aa89b0e10f08f23ee70cbe010a240b51bd620347b310fbe60a6","type":null,"confirmations":0}, receipt={"to":"0x748fA9975Afd2076BAc17B4223be5F3271cc9AC5","from":"0xE90E12261CCb0F3F7976Ae611A29e84a6A85f424","contractAddress":null,"transactionIndex":2,"root":"0x9ae51096a60f1704ed381854a01304833bae70d0069aa4312ead2c2d7ee84b28","gasUsed":{"type":"BigNumber","hex":"0x0940e9"},"logsBloom":"0x000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
<img width="972" alt="image" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/35927054/e952ad1a-d72b-4d1c-8eee-c252bb41d744">

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

https://github.com/Uniswap/v2-periphery/blob/master/contracts/examples/ExampleSwapToPrice.sol

### Can you provide the relevant portions of your contract code where the issue is occurring?

```
function swapToPrice(
        address tokenA,
        address tokenB,
        uint256 truePriceTokenA,
        uint256 truePriceTokenB,
        uint256 maxSpendTokenA,
        uint256 maxSpendTokenB,
        address to,
        uint256 deadline
    ) public {
        // true price is expressed as a ratio, so both values must be non-zero
        require(truePriceTokenA != 0 && truePriceTokenB != 0, "ExampleSwapToPrice: ZERO_PRICE");
        // caller can specify 0 for either if they wish to swap in only one direction, but not both
        require(maxSpendTokenA != 0 || maxSpendTokenB != 0, "ExampleSwapToPrice: ZERO_SPEND");

        bool aToB;
        uint256 amountIn;
        {
            (uint256 reserveA, uint256 reserveB) = ZksLibrary.getReserves(factory, tokenA, tokenB);
//            emit Log3(reserveA, reserveB);
            (aToB, amountIn) = ZksLiquidityMathLibrary.computeProfitMaximizingTrade(
                truePriceTokenA, truePriceTokenB,
                reserveA, reserveB
            );
//            emit Log3(reserveA, reserveB);
        }
//        emit Log3(reserveA, reserveB);
        require(amountIn > 0, 'ExampleSwapToPrice: ZERO_AMOUNT_IN');

        // spend up to the allowance of the token in
        uint256 maxSpend = aToB ? maxSpendTokenA : maxSpendTokenB;
        if (amountIn > maxSpend) {
            amountIn = maxSpend;
        }

        address tokenIn = aToB ? tokenA : tokenB;
        address tokenOut = aToB ? tokenB : tokenA;
        TransferHelper.safeTransferFrom(tokenIn, msg.sender, address(this), amountIn);
        TransferHelper.safeApprove(tokenIn, address(router), amountIn);

        address[] memory path = new address[](2);
        path[0] = tokenIn;
        path[1] = tokenOut;

        router.swapExactTokensForTokens(
            amountIn,
            0, // amountOutMin: we can skip computing this number because the math is tested
            path,
            to,
            deadline
        );
    }
```

### Have you tried to isolate the problem, and if so, what were the results?

Yes, I have tried and finally determined that the issue is with this line of code: `(uint256 reserveA, uint256 reserveB) = ZksLibrary.getReserves(factory, tokenA, tokenB);` However, the problem has not been resolved yet.

### What steps have you already taken to try to resolve the issue?

Define an `event Log3(uint256 indexed amountA, uint256 indexed amountB);` and change
```
(uint256 reserveA, uint256 reserveB) = ZksLibrary.getReserves(factory, tokenA, tokenB);
(aToB, amountIn) = ZksLiquidityMathLibrary.computeProfitMaximizingTrade(
  truePriceTokenA, truePriceTokenB,
  reserveA, reserveB
);
```
to
```
(uint256 reserveA, uint256 reserveB) = ZksLibrary.getReserves(factory, tokenA, tokenB);
emit Log3(reserveA, reserveB);
(aToB, amountIn) = ZksLiquidityMathLibrary.computeProfitMaximizingTrade(
  truePriceTokenA, truePriceTokenB,
  reserveA, reserveB
);
```
then the transaction will execute successfully.

### Additional Details

At the moment, this is as far as I could locate the problem. Do you have any insights or suggestions to share? Moreover, I still want to know how to debug this type of code. Currently, due to the uncertainty of how to properly debug this kind of problem, I can't pinpoint the issue effectively and accurately. Are there any reference documents available?

**Comment by bxpana** [comment]
@ThreeAndTwo 

Another user had run into an issue with a Uniswap V2 fork and this is what fixed it for them, let us know if that helps:

> The problem: Generally initial code hash is used in `pairFor` function in uniswap v2. Hardcoding the initial code hash won't cause an issue in EVM-based chains, but since the compiler of zkSync is different, it is different for zkSync side. So, modifying the init code hash **or** getting pair address by calling `getPair` function of factory contract are needed. Second way consumes a bit gas of course due to external call.  
https://github.com/Uniswap/v2-periphery/blob/master/contracts/libraries/UniswapV2Library.sol#L24

**Reply by ThreeAndTwo** [reply]
I don't think this will work because I made the changes according to the official documentation of zksync era. Moreover, the key point is that other cases are working fine. If there is an issue with create2, then all cases should fail, which is my current understanding. Therefore, I am desperate to know if there is a more efficient method for debugging.
<img width="1154" alt="image" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/35927054/88adc183-41d8-42e6-9636-09035f6875e6">
<img width="841" alt="image" src="https://github.com/zkSync-Community-Hub/zkync-developers/assets/35927054/ea9fc3be-fe52-427f-9b2f-f5dc57488869">

**Reply by IraChek** [reply]
someone should know about this

**Reply by MexicanAce** [reply, marked as answer]
Hi @ThreeAndTwo 👋 

One way to further debug issues like this is to use the new [In-Memory Node](https://github.com/matter-labs/era-test-node) and point your contract against a local node that prints out additional information.

Example (if you'd like to see the full stack of call traces for each transaction in a human readable format):
```bash
era_test_node --show-calls=all --resolve-hashes run
```

If you want to dig deeper, you can also test bootloader and system contracts ([documentation available here](https://era.zksync.io/docs/tools/testing/era-test-node.html#testing-bootloader-and-system-contracts))

**Comment by caydiverlarachel** [comment]
It seems like the issue might be related to the ZksLibrary.getReserves function not returning the expected values. You've tried debugging by emitting events, which is a good approach.

To further debug, you can:

Verify that the getReserves function is correctly implemented and returning the reserves for the specified tokens.
Ensure that the factory address is correctly set.
Check if there are any constraints or conditions in the getReserves function that might affect its behavior.
As for debugging techniques, emitting events and using console logs are effective methods. You can also use tools like Truffle Debugger or Hardhat Console for more advanced debugging. Additionally, reviewing the documentation and source code of the libraries you're using can provide valuable insights into their behavior and potential issues.

**Comment by cahillnick** [comment]
It seems like the issue might be related to the getReserves function in the ZksLibrary contract. Ensure that this function returns the correct values for reserves. Also, verify that the computeProfitMaximizingTrade function is receiving the correct input parameters.

To debug this type of code, you can:

Use events: Emit events at different stages of the execution to track the flow and values of variables.
Check input parameters: Ensure that all input parameters passed to functions are correct and valid.
Review contracts: Double-check the implementation of the getReserves and computeProfitMaximizingTrade functions for any errors.
Test locally: Use a local development environment to debug and test the contract code more efficiently.
Consult documentation: Refer to the documentation of the contracts and libraries being used for any specific debugging guidelines or tips.

**Comment by dariuscandfield** [comment]
It seems like there's an issue with the `getReserves` function call from the zkSwap library. Ensure that the parameters passed to this function are correct and that the library is functioning as expected. Additionally, consider debugging the code by emitting events or logging intermediate values to better understand the flow of execution and identify potential issues. Refer to zkSwap documentation or seek assistance from their community for further debugging guidance.
